// stylable/dist/src/optimizer/classname-optimizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var selector_utils_1 = require("../selector-utils");
var StylableClassNameOptimizer = /** @class */ (function () {
    function StylableClassNameOptimizer() {
        this.context = {
            names: {}
        };
    }
    StylableClassNameOptimizer.prototype.rewriteSelector = function (selector) {
        var _this = this;
        var ast = selector_utils_1.parseSelector(selector);
        selector_utils_1.traverseNode(ast, function (node) {
            if (node.type === 'class') {
                if (!_this.context.names[node.name]) {
                    _this.generateName(node.name);
                }
                node.name = _this.context.names[node.name];
            }
        });
        return selector_utils_1.stringifySelector(ast);
    };
    StylableClassNameOptimizer.prototype.generateName = function (name) {
        return (this.context.names[name] = 's' + Object.keys(this.context.names).length);
    };
    StylableClassNameOptimizer.prototype.optimizeAstAndExports = function (ast, exported, classes) {
        var _this = this;
        if (classes === void 0) { classes = Object.keys(exported); }
        ast.walkRules(function (rule) {
            rule.selector = _this.rewriteSelector(rule.selector);
        });
        classes.forEach(function (originName) {
            if (exported[originName]) {
                exported[originName] = exported[originName]
                    .split(' ')
                    .map(function (renderedNamed) {
                    return (_this.context.names[renderedNamed] || _this.generateName(renderedNamed));
                })
                    .join(' ');
            }
        });
    };
    return StylableClassNameOptimizer;
}());
exports.StylableClassNameOptimizer = StylableClassNameOptimizer;
