// bit-bin/dist/extensions/flows/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FlowsExt", {
  enumerable: true,
  get: function () {
    return _flows().default;
  }
});
Object.defineProperty(exports, "Flows", {
  enumerable: true,
  get: function () {
    return _flows2().Flows;
  }
});
Object.defineProperty(exports, "IdsAndFlows", {
  enumerable: true,
  get: function () {
    return _flows2().IdsAndFlows;
  }
});
Object.defineProperty(exports, "TASK_SEPARATOR", {
  enumerable: true,
  get: function () {
    return _task().TASK_SEPARATOR;
  }
});
Object.defineProperty(exports, "SCRIPT_FILENAME", {
  enumerable: true,
  get: function () {
    return _task().SCRIPT_FILENAME;
  }
});
Object.defineProperty(exports, "flattenNestedMap", {
  enumerable: true,
  get: function () {
    return _flattenNestedMap().flattenNestedMap;
  }
});
Object.defineProperty(exports, "flattenReplaySubject", {
  enumerable: true,
  get: function () {
    return _flattenNestedMap().flattenReplaySubject;
  }
});
Object.defineProperty(exports, "createFakeCapsule", {
  enumerable: true,
  get: function () {
    return _createCapsule().createFakeCapsule;
  }
});
Object.defineProperty(exports, "createGetGraphFn", {
  enumerable: true,
  get: function () {
    return _createFakeNetwork().createGetGraphFn;
  }
});
Object.defineProperty(exports, "createTestNetworkStream", {
  enumerable: true,
  get: function () {
    return _createFakeNetwork().createTestNetworkStream;
  }
});

function _flows() {
  const data = _interopRequireDefault(require("./flows.manifest"));

  _flows = function () {
    return data;
  };

  return data;
}

function _flows2() {
  const data = require("./flows");

  _flows2 = function () {
    return data;
  };

  return data;
}

function _task() {
  const data = require("./task/task");

  _task = function () {
    return data;
  };

  return data;
}

function _flattenNestedMap() {
  const data = require("./util/flatten-nested-map");

  _flattenNestedMap = function () {
    return data;
  };

  return data;
}

function _createCapsule() {
  const data = require("./util/create-capsule");

  _createCapsule = function () {
    return data;
  };

  return data;
}

function _createFakeNetwork() {
  const data = require("./util/create-fake-network");

  _createFakeNetwork = function () {
    return data;
  };

  return data;
}
