// bit-bin/dist/extensions/cli/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  CLIExtension: true,
  Command: true,
  CLIArgs: true,
  Flags: true,
  GenericObject: true,
  CommandOptions: true,
  getID: true
};
Object.defineProperty(exports, "CLIExtension", {
  enumerable: true,
  get: function () {
    return _cli().CLIExtension;
  }
});
Object.defineProperty(exports, "Command", {
  enumerable: true,
  get: function () {
    return _command().Command;
  }
});
Object.defineProperty(exports, "CLIArgs", {
  enumerable: true,
  get: function () {
    return _command().CLIArgs;
  }
});
Object.defineProperty(exports, "Flags", {
  enumerable: true,
  get: function () {
    return _command().Flags;
  }
});
Object.defineProperty(exports, "GenericObject", {
  enumerable: true,
  get: function () {
    return _command().GenericObject;
  }
});
Object.defineProperty(exports, "CommandOptions", {
  enumerable: true,
  get: function () {
    return _legacyCommand().CommandOptions;
  }
});
Object.defineProperty(exports, "getID", {
  enumerable: true,
  get: function () {
    return _registry().getID;
  }
});

function _cli() {
  const data = require("./cli.extension");

  _cli = function () {
    return data;
  };

  return data;
}

function _command() {
  const data = require("../../cli/command");

  _command = function () {
    return data;
  };

  return data;
}

function _legacyCommand() {
  const data = require("../../cli/legacy-command");

  _legacyCommand = function () {
    return data;
  };

  return data;
}

function _registry() {
  const data = require("./registry");

  _registry = function () {
    return data;
  };

  return data;
}

var _exceptions = require("./exceptions");

Object.keys(_exceptions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _exceptions[key];
    }
  });
});
