// bit-bin/dist/consumer/config/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LegacyWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().default;
  }
});
Object.defineProperty(exports, "LegacyWorkspaceConfigProps", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().WorkspaceConfigProps;
  }
});
Object.defineProperty(exports, "WorkspaceConfigEnsureFunction", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().WorkspaceConfigEnsureFunction;
  }
});
Object.defineProperty(exports, "WorkspaceConfigLoadFunction", {
  enumerable: true,
  get: function () {
    return _workspaceConfig().WorkspaceConfigLoadFunction;
  }
});
Object.defineProperty(exports, "ILegacyWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _legacyWorkspaceConfigInterface().ILegacyWorkspaceConfig;
  }
});
Object.defineProperty(exports, "ExtensionConfigList", {
  enumerable: true,
  get: function () {
    return _extensionConfigList().ExtensionConfigList;
  }
});
Object.defineProperty(exports, "ExtensionConfigEntry", {
  enumerable: true,
  get: function () {
    return _extensionConfigList().ExtensionConfigEntry;
  }
});
Object.defineProperty(exports, "IExtensionConfigList", {
  enumerable: true,
  get: function () {
    return _extensionConfigList().IExtensionConfigList;
  }
});
exports.default = void 0;

function _componentConfig() {
  const data = _interopRequireDefault(require("./component-config"));

  _componentConfig = function () {
    return data;
  };

  return data;
}

function _workspaceConfig() {
  const data = _interopRequireWildcard(require("./workspace-config"));

  _workspaceConfig = function () {
    return data;
  };

  return data;
}

function _legacyWorkspaceConfigInterface() {
  const data = require("./legacy-workspace-config-interface");

  _legacyWorkspaceConfigInterface = function () {
    return data;
  };

  return data;
}

function _extensionConfigList() {
  const data = require("./extension-config-list");

  _extensionConfigList = function () {
    return data;
  };

  return data;
}

var _default = _componentConfig().default;

exports.default = _default;
