// bit-bin/dist/extensions/create/create.cmd.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateCmd = void 0;

function _bluebird() {
  const data = require("bluebird");

  _bluebird = function () {
    return data;
  };

  return data;
}

function _defineProperty2() {
  const data = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

  _defineProperty2 = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _ink() {
  const data = require("ink");

  _ink = function () {
    return data;
  };

  return data;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
// eslint-disable-next-line @typescript-eslint/no-unused-vars
class CreateCmd {
  constructor(create) {
    this.create = create;
    (0, _defineProperty2().default)(this, "name", 'create [name]');
    (0, _defineProperty2().default)(this, "description", 'create a new component from a template');
    (0, _defineProperty2().default)(this, "shortDescription", '');
    (0, _defineProperty2().default)(this, "alias", '');
    (0, _defineProperty2().default)(this, "group", '');
    (0, _defineProperty2().default)(this, "private", true);
    (0, _defineProperty2().default)(this, "options", []);
  }

  render([name]) {
    var _this = this;

    return (0, _bluebird().coroutine)(function* () {
      // @ts-ignore
      const results = yield _this.create.create(name);
      const result = results.addedComponents[0];
      const files = result.files.map((item, key) => /*#__PURE__*/_react().default.createElement("li", {
        key: key
      }, /*#__PURE__*/_react().default.createElement(_ink().Color, {
        green: true
      }, item.relativePath))); // eslint-disable-next-line @typescript-eslint/no-unused-vars

      const AddedComponent = () => /*#__PURE__*/_react().default.createElement(_ink().Box, {
        padding: 1,
        flexDirection: "column"
      }, /*#__PURE__*/_react().default.createElement(_ink().Box, null, "tracking component ", /*#__PURE__*/_react().default.createElement(_ink().Text, {
        bold: true
      }, result.id)), /*#__PURE__*/_react().default.createElement(_ink().Box, {
        paddingLeft: 2
      }, /*#__PURE__*/_react().default.createElement("ul", null, files)));

      return /*#__PURE__*/_react().default.createElement(AddedComponent, null);
    })();
  }

}

exports.CreateCmd = CreateCmd;
