// bit-bin/dist/extensions/dependency-resolver/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DependencyResolverExtension", {
  enumerable: true,
  get: function () {
    return _dependencyResolver().DependencyResolverExtension;
  }
});
Object.defineProperty(exports, "DependenciesDefinition", {
  enumerable: true,
  get: function () {
    return _types().DependenciesDefinition;
  }
});
Object.defineProperty(exports, "SemverVersion", {
  enumerable: true,
  get: function () {
    return _types().SemverVersion;
  }
});
Object.defineProperty(exports, "SemverVersionRule", {
  enumerable: true,
  get: function () {
    return _types().SemverVersionRule;
  }
});
Object.defineProperty(exports, "DependencyResolverWorkspaceConfig", {
  enumerable: true,
  get: function () {
    return _types().DependencyResolverWorkspaceConfig;
  }
});
Object.defineProperty(exports, "DependencyResolverVariantConfig", {
  enumerable: true,
  get: function () {
    return _types().DependencyResolverVariantConfig;
  }
});
Object.defineProperty(exports, "RawComponentState", {
  enumerable: true,
  get: function () {
    return _types().RawComponentState;
  }
});
Object.defineProperty(exports, "DependencyType", {
  enumerable: true,
  get: function () {
    return _types().DependencyType;
  }
});

function _dependencyResolver() {
  const data = require("./dependency-resolver.extension");

  _dependencyResolver = function () {
    return data;
  };

  return data;
}

function _types() {
  const data = require("./types");

  _types = function () {
    return data;
  };

  return data;
}
