// bit-bin/dist/extensions/graph/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ComponentGraph", {
  enumerable: true,
  get: function () {
    return _componentGraph().ComponentGraph;
  }
});
Object.defineProperty(exports, "GraphBuilder", {
  enumerable: true,
  get: function () {
    return _graphBuilder().GraphBuilder;
  }
});
Object.defineProperty(exports, "ComponentGraphExt", {
  enumerable: true,
  get: function () {
    return _graph().default;
  }
});
Object.defineProperty(exports, "Dependency", {
  enumerable: true,
  get: function () {
    return _dependency().Dependency;
  }
});
Object.defineProperty(exports, "DuplicateDependency", {
  enumerable: true,
  get: function () {
    return _duplicateDependency().DuplicateDependency;
  }
});
Object.defineProperty(exports, "VersionSubgraph", {
  enumerable: true,
  get: function () {
    return _duplicateDependency().VersionSubgraph;
  }
});

function _componentGraph() {
  const data = require("./component-graph");

  _componentGraph = function () {
    return data;
  };

  return data;
}

function _graphBuilder() {
  const data = require("./graph-builder");

  _graphBuilder = function () {
    return data;
  };

  return data;
}

function _graph() {
  const data = _interopRequireDefault(require("./graph.manifest"));

  _graph = function () {
    return data;
  };

  return data;
}

function _dependency() {
  const data = require("./dependency");

  _dependency = function () {
    return data;
  };

  return data;
}

function _duplicateDependency() {
  const data = require("./duplicate-dependency");

  _duplicateDependency = function () {
    return data;
  };

  return data;
}
