// bit-bin/dist/extensions/component/id.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _bitId() {
  const data = require("../../bit-id");

  _bitId = function () {
    return data;
  };

  return data;
}

class ComponentID {
  constructor(legacyComponentId) {
    this.legacyComponentId = legacyComponentId;
  }
  /**
   * An access to the legacy id. DO NOT USE THIS
   *
   * @readonly
   * @memberof ComponentID
   */


  get _legacy() {
    return this.legacyComponentId;
  }
  /**
   * resolves the version of the component ID.
   */


  get version() {
    return this.legacyComponentId.version;
  }
  /**
   * resolves the name of the component.
   */


  get name() {
    return this.legacyComponentId.name;
  }

  isEqual(id) {
    return this._legacy.isEqual(id._legacy);
  }
  /**
   * serialize the component ID.
   */


  toString() {
    return this.legacyComponentId.toString();
  }
  /**
   * generate a component ID from a string.
   */


  static fromString(idStr) {
    return new ComponentID(_bitId().BitId.parse(idStr));
  }
  /**
   * create a `ComponentID` instance from the legacy `BitId`.
   */


  static fromLegacy(legacyId) {
    return new ComponentID(legacyId);
  }

}

exports.default = ComponentID;
