// bit-bin/dist/extensions/component/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Component", {
  enumerable: true,
  get: function () {
    return _component().default;
  }
});
Object.defineProperty(exports, "ComponentFactoryExt", {
  enumerable: true,
  get: function () {
    return _componentFactory().default;
  }
});
Object.defineProperty(exports, "ComponentFactory", {
  enumerable: true,
  get: function () {
    return _componentFactory().default;
  }
});
Object.defineProperty(exports, "ComponentID", {
  enumerable: true,
  get: function () {
    return _id().default;
  }
});
Object.defineProperty(exports, "ComponentFS", {
  enumerable: true,
  get: function () {
    return _componentFs().default;
  }
});
Object.defineProperty(exports, "ComponentConfig", {
  enumerable: true,
  get: function () {
    return _config().default;
  }
});

function _component() {
  const data = _interopRequireDefault(require("./component"));

  _component = function () {
    return data;
  };

  return data;
}

function _componentFactory() {
  const data = _interopRequireDefault(require("./component-factory.extension"));

  _componentFactory = function () {
    return data;
  };

  return data;
}

function _id() {
  const data = _interopRequireDefault(require("./id"));

  _id = function () {
    return data;
  };

  return data;
}

function _componentFs() {
  const data = _interopRequireDefault(require("./component-fs"));

  _componentFs = function () {
    return data;
  };

  return data;
}

function _config() {
  const data = _interopRequireDefault(require("./config"));

  _config = function () {
    return data;
  };

  return data;
}
