// bit-bin/dist/extensions/component/state.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _config() {
  const data = _interopRequireDefault(require("./config"));

  _config = function () {
    return data;
  };

  return data;
}

function _store() {
  const data = require("./store");

  _store = function () {
    return data;
  };

  return data;
}

function _componentFs() {
  const data = _interopRequireDefault(require("./component-fs"));

  _componentFs = function () {
    return data;
  };

  return data;
}

class State {
  constructor(config, filesystem, store, dependencies, _consumer) {
    this.config = config;
    this.filesystem = filesystem;
    this.store = store;
    this.dependencies = dependencies;
    this._consumer = _consumer;
  }
  /**
   * calculate the hash of this state
   */


  get hash() {
    return '';
  }

  dependencyGraph() {}

  static fromLegacy(consumerComponent) {
    return new State(new (_config().default)(consumerComponent.mainFile, consumerComponent.extensions), _componentFs().default.fromVinyls(consumerComponent.files), _store().Store.fromArray([]), consumerComponent.dependencies, consumerComponent);
  }

}

exports.default = State;
