// stylable/dist/src/cached-process-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function cachedProcessFile(processor, fs, resolvePath) {
    var cache = {};
    var postProcessors = [];
    function process(fullpath, ignoreCache, context) {
        if (ignoreCache === void 0) { ignoreCache = false; }
        var resolvedPath = resolvePath(fullpath, context);
        var stat = fs.statSync(resolvedPath);
        var cached = cache[resolvedPath];
        if (ignoreCache ||
            !cached ||
            (cached && cached.stat.mtime.valueOf() !== stat.mtime.valueOf())) {
            var content = fs.readFileSync(resolvedPath, 'utf8');
            var value = processContent(content, resolvedPath);
            cache[resolvedPath] = { value: value, stat: { mtime: stat.mtime } };
        }
        return cache[resolvedPath].value;
    }
    function processContent(content, filePath) {
        return postProcessors.reduce(function (value, postProcessor) {
            return postProcessor(value, filePath);
        }, processor(filePath, content));
    }
    function add(fullpath, value) {
        var mtime;
        try {
            mtime = fs.statSync(fullpath).mtime;
        }
        catch (e) {
            mtime = new Date();
        }
        cache[fullpath] = {
            value: value,
            stat: {
                mtime: mtime
            }
        };
    }
    return {
        processContent: processContent,
        postProcessors: postProcessors,
        cache: cache,
        process: process,
        add: add
    };
}
exports.cachedProcessFile = cachedProcessFile;
