// stylable/dist/src/memory-minimal-fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var deindent = require('deindent');
var path_1 = require("./path");
function createMinimalFS(config) {
    var files = config.files;
    for (var file in files) {
        if (files[file].mtime === undefined) {
            files[file].mtime = new Date();
            files[path_1.resolve(file)] = files[file];
        }
    }
    function isDir(path) {
        return Object.keys(files).some(function (p) {
            return p.startsWith(path[path.length - 1] === path_1.sep ? path : path + path_1.sep);
        });
    }
    var fs = {
        readFileSync: function (path) {
            if (!files[path]) {
                throw new Error('Cannot find file: ' + path);
            }
            if (config.trimWS) {
                return deindent(files[path].content).trim();
            }
            return files[path].content;
        },
        statSync: function (path) {
            var isDirectory = isDir(path);
            if (!files[path] && !isDirectory) {
                throw new Error('Cannot find file: ' + path);
            }
            return {
                isDirectory: function () {
                    return isDirectory;
                },
                isFile: function () {
                    return true;
                },
                mtime: isDirectory ? new Date() : files[path].mtime
            };
        }
    };
    var requireModule = function require(id) {
        var _module = {
            id: id,
            exports: {}
        };
        try {
            if (!id.match(/\.js$/)) {
                id += '.js';
            }
            var fn = new Function('module', 'exports', 'require', files[id].content);
            fn(_module, _module.exports, requireModule);
        }
        catch (e) {
            throw new Error('Cannot require file: ' + id);
        }
        return _module.exports;
    };
    function resolvePath(_ctx, path) {
        return path;
    }
    return {
        fs: fs,
        requireModule: requireModule,
        resolvePath: resolvePath
    };
}
exports.createMinimalFS = createMinimalFS;
