// stylable/dist/src/path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var sPath = {
    resolve: path.resolve,
    isAbsolute: path.isAbsolute,
    join: path.join,
    basename: path.basename,
    dirname: path.dirname,
    relative: path.relative,
    sep: path.sep
};
exports.default = sPath;
exports.sep = {
    toString: function () {
        return sPath.sep;
    }
};
exports.resolve = function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return sPath.resolve.apply(sPath, args);
};
exports.isAbsolute = function (path) {
    return sPath.isAbsolute(path);
};
exports.join = function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return sPath.join.apply(sPath, args);
};
exports.basename = function (p, ext) {
    return sPath.basename(p, ext);
};
exports.dirname = function (p) {
    return sPath.dirname(p);
};
exports.relative = function (from, to) {
    return sPath.relative(from, to);
};
