// stylable/dist/src/stylable-meta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var stylable_utils_1 = require("./stylable-utils");
var stylable_value_parsers_1 = require("./stylable-value-parsers");
exports.RESERVED_ROOT_NAME = 'root';
var StylableMeta = /** @class */ (function () {
    function StylableMeta(ast, diagnostics) {
        var _a, _b, _c;
        this.ast = ast;
        this.diagnostics = diagnostics;
        var rootSymbol = (_a = {
                _kind: 'class',
                name: exports.RESERVED_ROOT_NAME
            },
            _a[stylable_value_parsers_1.valueMapping.root] = true,
            _a);
        this.rawAst = ast.clone();
        this.source = stylable_utils_1.getSourcePath(ast, diagnostics);
        this.root = exports.RESERVED_ROOT_NAME;
        this.namespace = '';
        this.imports = [];
        this.vars = [];
        this.keyframes = [];
        this.elements = {};
        this.classes = (_b = {},
            _b[exports.RESERVED_ROOT_NAME] = rootSymbol,
            _b);
        this.mappedSymbols = (_c = {},
            _c[exports.RESERVED_ROOT_NAME] = rootSymbol,
            _c);
        this.customSelectors = {};
        this.urls = [];
        this.transformDiagnostics = null;
    }
    return StylableMeta;
}());
exports.StylableMeta = StylableMeta;
