// bit-bin/dist/extensions/graphql/graphql.ui.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphQlUI = void 0;

function _bluebird() {
  const data = require("bluebird");

  _bluebird = function () {
    return data;
  };

  return data;
}

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function () {
    return data;
  };

  return data;
}

function _apolloBoost() {
  const data = _interopRequireDefault(require("apollo-boost"));

  _apolloBoost = function () {
    return data;
  };

  return data;
}

function _graphqlProvider() {
  const data = require("./graphql-provider");

  _graphqlProvider = function () {
    return data;
  };

  return data;
}

class GraphQlUI {
  constructor(client) {
    this.client = client;
  }
  /**
   * get the graphQL provider
   */


  getProvider(rootComponent) {
    return /*#__PURE__*/_react().default.createElement(_graphqlProvider().GraphQLProvider, {
      client: this.client,
      root: rootComponent
    });
  }

  static provider() {
    return (0, _bluebird().coroutine)(function* () {
      const client = new (_apolloBoost().default)({
        uri: 'http://localhost:4000/graphql'
      });
      return new GraphQlUI(client);
    })();
  }

}

exports.GraphQlUI = GraphQlUI;
